File: backend/src/modules/auth/auth.service.ts
import {
  ConflictException,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { JwtService } from '@nestjs/jwt';
import { createHash, randomBytes } from 'crypto';
import * as argon2 from 'argon2';
import { PrismaService } from '../../prisma/prisma.service';
import { LoginDto } from './dto/login.dto';
import { RefreshTokenDto } from './dto/refresh-token.dto';
import { RegisterDto } from './dto/register.dto';

@Injectable()
export class AuthService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly jwt: JwtService,
    private readonly config: ConfigService,
  ) {}

  async register(dto: RegisterDto) {
    const passwordHash = await argon2.hash(dto.password);
    try {
      const user = await this.prisma.user.create({
        data: {
          username: dto.username.trim(),
          email: dto.email?.trim().toLowerCase() || null,
          passwordHash,
          profile: {
            create: {
              nickname: dto.nickname?.trim() || '学习者',
            },
          },
        },
        include: { profile: true },
      });

      return this.issueTokenPair(user.id, user.username, this.serializeUser(user));
    } catch (error) {
      if (this.isUniqueError(error)) {
        throw new ConflictException('用户名或邮箱已被注册');
      }
      throw error;
    }
  }

  async login(dto: LoginDto) {
    const identifier = dto.identifier.trim();
    const user = await this.prisma.user.findFirst({
      where: {
        OR: [
          { username: identifier },
          { email: identifier.toLowerCase() },
        ],
      },
      include: { profile: true },
    });

    if (!user || !(await argon2.verify(user.passwordHash, dto.password))) {
      throw new UnauthorizedException('账号或密码错误');
    }

    return this.issueTokenPair(user.id, user.username, this.serializeUser(user));
  }

  async refresh(dto: RefreshTokenDto) {
    const tokenHash = this.hashToken(dto.refreshToken);
    const saved = await this.prisma.refreshToken.findUnique({
      where: { tokenHash },
      include: { user: { include: { profile: true } } },
    });

    if (!saved || saved.revokedAt || saved.expiresAt.getTime() <= Date.now()) {
      throw new UnauthorizedException('刷新凭据无效或已过期');
    }

    await this.prisma.refreshToken.update({
      where: { id: saved.id },
      data: { revokedAt: new Date() },
    });

    return this.issueTokenPair(
      saved.user.id,
      saved.user.username,
      this.serializeUser(saved.user),
    );
  }

  async logout(userId: string, refreshToken?: string) {
    if (refreshToken) {
      await this.prisma.refreshToken.updateMany({
        where: {
          userId,
          tokenHash: this.hashToken(refreshToken),
          revokedAt: null,
        },
        data: { revokedAt: new Date() },
      });
      return { ok: true };
    }
    await this.prisma.refreshToken.updateMany({
      where: { userId, revokedAt: null },
      data: { revokedAt: new Date() },
    });
    return { ok: true };
  }

  async deleteAccount(userId: string, accessToken?: string) {
    await this.prisma.$transaction([
      this.prisma.refreshToken.updateMany({
        where: { userId, revokedAt: null },
        data: { revokedAt: new Date() },
      }),
      this.prisma.user.delete({ where: { id: userId } }),
    ]);
    void accessToken;
    return { ok: true };
  }

  private async issueTokenPair(userId: string, username: string, user: unknown) {
    const accessToken = await this.jwt.signAsync({ sub: userId, username });
    const refreshToken = randomBytes(48).toString('base64url');
    const days = Number(this.config.get('REFRESH_TOKEN_DAYS') ?? 30);
    const expiresAt = new Date(Date.now() + days * 24 * 60 * 60 * 1000);

    await this.prisma.refreshToken.create({
      data: {
        userId,
        tokenHash: this.hashToken(refreshToken),
        expiresAt,
      },
    });

    return { user, accessToken, refreshToken };
  }

  private hashToken(token: string) {
    return createHash('sha256').update(token).digest('hex');
  }

  private serializeUser(user: {
    id: string;
    username: string;
    email: string | null;
    createdAt: Date;
    profile?: unknown;
  }) {
    return {
      id: user.id,
      username: user.username,
      email: user.email,
      profile: user.profile,
      createdAt: user.createdAt,
    };
  }

  private isUniqueError(error: unknown) {
    return typeof error === 'object' && error !== null && (error as { code?: string }).code === 'P2002';
  }
}

File: lib/src/services/auth_service.dart
import 'api_client.dart';

class AuthResult {
  const AuthResult({
    required this.accessToken,
    required this.refreshToken,
    this.user,
  });

  final String accessToken;
  final String refreshToken;
  final Map<String, dynamic>? user;
}

class AuthService {
  AuthService({ApiClient? apiClient}) : _api = apiClient;

  ApiClient? _api;

  ApiClient get api {
    final client = _api;
    if (client == null) {
      throw const ApiException('尚未初始化后端连接，请先配置服务地址');
    }
    return client;
  }

  void attach(ApiClient client) {
    _api = client;
  }

  Future<AuthResult> register({
    required String username,
    required String password,
    String? nickname,
  }) async {
    final data = await api.postJson('/auth/register', body: {
      'username': username,
      'password': password,
      if (nickname != null && nickname.trim().isNotEmpty)
        'nickname': nickname.trim(),
    });
    return _parseAuthResult(data);
  }

  Future<AuthResult> login({
    required String identifier,
    required String password,
  }) async {
    final data = await api.postJson('/auth/login', body: {
      'identifier': identifier,
      'password': password,
    });
    return _parseAuthResult(data);
  }

  Future<AuthResult> refreshToken(String refreshToken) async {
    final data = await api.postJson('/auth/refresh', body: {
      'refreshToken': refreshToken,
    });
    return _parseAuthResult(data);
  }

  Future<Map<String, dynamic>> getProfile() async {
    return api.getJson('/me');
  }

  Future<Map<String, dynamic>> updateProfile({
    String? nickname,
    String? avatarEmoji,
    String? avatarImageUrl,
    String? bio,
  }) async {
    final body = <String, dynamic>{};
    if (nickname != null) body['nickname'] = nickname;
    if (avatarEmoji != null) body['avatarEmoji'] = avatarEmoji;
    if (avatarImageUrl != null) body['avatarImageUrl'] = avatarImageUrl;
    if (bio != null) body['bio'] = bio;
    return api.patchJson('/me/profile', body: body);
  }

  Future<void> logout({String? refreshToken}) async {
    await api.postJson('/auth/logout', body: {
      if (refreshToken != null && refreshToken.isNotEmpty)
        'refreshToken': refreshToken,
    });
  }

  Future<void> deleteAccount() async {
    await api.postJson('/auth/delete-account');
  }

  AuthResult _parseAuthResult(Map<String, dynamic> data) {
    final accessToken = data['accessToken'] as String?;
    final refreshToken = data['refreshToken'] as String?;
    if (accessToken == null || accessToken.isEmpty) {
      throw const ApiException('服务器未返回登录凭据');
    }
    if (refreshToken == null || refreshToken.isEmpty) {
      throw const ApiException('服务器未返回刷新凭据');
    }
    return AuthResult(
      accessToken: accessToken,
      refreshToken: refreshToken,
      user: data['user'] is Map<String, dynamic>
          ? data['user'] as Map<String, dynamic>
          : null,
    );
  }
}
